src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs

```diff
@@ -161,6 +169,46 @@ public int InitializeCommunication()
         return endpoint.GetIpEndPoint().Port;
     }
 
+
+    private bool TrySetupMessageReceiver(
+        EventHandler<MessageReceivedEventArgs> onMessageReceived,
+        Action<DisconnectedEventArgs> onDisconnected)
+    {
+        TPDebug.Assert(_channel is not null, "_channel is null");
+
+        // Note: Attempts to setup a message receiver.
+        // It's possible that the testhost was already disconnected and in that case we should
+        // immediately call the disconnected callback.
+
+        // Design: The current method is needed because the request sender sets up
+        //         the disconnect handler late. If the first thing that is done by the class
+        //         is to setup the disconnect handler, then we'd only need to fire the handler
+        //         when the disconnect event fires.
+
+        _onDisconnected = onDisconnected;
+
+        // If the testhost was already disconnected, trigger the handler immediately.
+        if (_disconnectedInfo is DisconnectedEventArgs args)
+        {
+            InvokeDisconnectedHandler(args);
+            return false;
+        }
+
+        _onMessageReceived = onMessageReceived;
+        _channel.MessageReceived += _onMessageReceived;
+
+        return true;
+    }
+
+    private void InvokeDisconnectedHandler(DisconnectedEventArgs args)
```

Comment:
❔ I am wondering if there is a real benefit in taking that parameter instead of simply using the field from the body.

Reply:
this is a safety against any potential threading changes in the future, like if the code would be changed to set _disconnectedInfo to null if the handler is called.

Reply:
I am not super big fan of planning future of things we don't know how will evolve because my experience proves me it's never going as planned but I guess that's ok here.